src/util/colorize.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+Part of Minetest
+Copyright (C) 2024 rubenwardy
+
+Permission to use, copy, modify, and distribute this software for any
+purpose with or without fee is hereby granted, provided that the above
+copyright notice and this permission notice appear in all copies.
+
+THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+
+#include "config.h"
+
+#ifdef USE_CURL
+#include <curl/urlapi.h>
+#endif
+
+#include "colorize.h"
+#include "log.h"
+#include <sstream>
+
+std::string colorize_url(const std::string &url) {
```

Comment:
api suggestion that would be less weird and cleaner IMO:
- have the function split the url into three parts `prefix, hostname, suffix`
- the caller can then colorize them as wanted

Reply:
This wouldn't allow colorising the escape codes red or returning error messages, what about:

```c++
/**
 * Colorize URL to highlight domain and unsafe characters
 *
 * @param out either the output string or an error message
 * @param url URL to format
 * @return true for success, false otherwise
 */
bool colorize_url(std::string &out, const std::string &url);
```

Reply:
sounds ok.